File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        if (n == 1) {
            return 1;
        } else if (n == 2) {
            return 2;
        }
        
        int x = 1;
        int y = 2;
        
        for (int i = 3; i <= n; i++) {
            int z = x + y;
            x = y;
            y = z;
            
        }
        
        return y;
    }
}

File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    def __init__(self):
        self.cache = {}
    def climbStairs(self, n: int) -> int:
        # dp = [0] * (n + 1)

        # dp[0] = 1
        # dp[1] = 1

        # for i in range(2, n + 1):
        #     dp[i] = dp[i - 1] + dp[i - 2]

        # return dp[n]
        
        if n == 0 or n == 1:
            return 1

        if n in self.cache:
            return self.cache[n]


        res =  self.climbStairs(n - 1) + self.climbStairs(n - 2)
        self.cache[n] = res
        return res